src/librustc/ty/outlives.rs

```diff
@@ -174,6 +174,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
             ty::TyEnum(..) |        // OutlivesNominalType
             ty::TyStruct(..) |      // OutlivesNominalType
             ty::TyBox(..) |         // OutlivesNominalType (ish)
+            ty::TyAnon(..) |        // OutlivesNominalType (ish)
```

Comment:
*Probably* this code is doing the right thing, but it's a bit more conservative than it might otherwise be. Really there are choices here just like for projections. Consider for example if you had `trait Foo<'a,'b>: 'a`, then you could in theory derive that `impl Foo<'a,'b>: 'a` even if `'b: 'a` doesn't necessarily hold. Something similar could occur with `trait Foo<'a, T>: 'a`, where `impl Foo<'a, &'b T>: 'a` might hold even if `'b: 'a` does not hold. 

It seems in a way like an odd rule, since the type name that the type system gives includes lifetimes (like `'b`) that are not necessarily in scope, but the same thing can occur with projections -- the point is that, after monomorphization, the normalized form of that type *is* known to outlive `'a`, even if all components pre-normalization do not.

/me has that itchy feeling of wanting to revisit a richer, formalized version of this, to try and have a good reference. But nothing here makes me too scared yet. =)